File: app/src/main/java/com/peterlaurence/trekme/ui/mapimport/MapImportFragment.java
package com.peterlaurence.trekme.ui.mapimport;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.peterlaurence.trekme.R;
import com.peterlaurence.trekme.core.events.MapArchiveListUpdateEvent;
import com.peterlaurence.trekme.core.map.maploader.MapLoader;
import com.peterlaurence.trekme.ui.events.DrawerClosedEvent;
import com.peterlaurence.trekme.ui.events.MapImportedEvent;
import com.peterlaurence.trekme.ui.events.RequestImportMapEvent;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

/**
 * A {@link Fragment} subclass that displays the list of maps archives available for import.
 *
 * @author peterLaurence on 08/06/16.
 */
public class MapImportFragment extends Fragment {

    private static String CREATE_FROM_SCREEN_ROTATE = "create";
    private RecyclerView mRecyclerView;
    private MapArchiveAdapter mMapArchiveAdapter;
    private OnMapArchiveFragmentInteractionListener mListener;
    private View mView;
    private boolean mCreateAfterScreenRotation = false;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        if (context instanceof OnMapArchiveFragmentInteractionListener) {
            mListener = (OnMapArchiveFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnMapArchiveFragmentInteractionListener");
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setHasOptionsMenu(false);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        return inflater.inflate(R.layout.fragment_map_import, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mView = view;

        /* When this fragment is created from a screen rotating, don't wait the drawer layout to
         * close to re-generate the map list.
         */
        if (savedInstanceState != null) {
            if (savedInstanceState.getBoolean(CREATE_FROM_SCREEN_ROTATE)) {
                mCreateAfterScreenRotation = true;
            }
        }

        mRecyclerView = view.findViewById(R.id.recyclerview_map_import);
        mRecyclerView.setHasFixedSize(false);

        LinearLayoutManager llm = new LinearLayoutManager(getContext());
        mRecyclerView.setLayoutManager(llm);

        mMapArchiveAdapter = new MapArchiveAdapter();
        mRecyclerView.setAdapter(mMapArchiveAdapter);

        /* Item decoration : divider */
        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(view.getContext(),
                DividerItemDecoration.VERTICAL);
        Drawable divider = view.getContext().getDrawable(R.drawable.divider);
        if (divider != null) {
            dividerItemDecoration.setDrawable(divider);
        }
        mRecyclerView.addItemDecoration(dividerItemDecoration);
    }

    @Override
    public void onStart() {
        super.onStart();
        EventBus.getDefault().register(this);
        if (mMapArchiveAdapter != null) {
            mMapArchiveAdapter.subscribeEventBus();
        }
        /* Request archive generation after EventBus registration to be 100% sure to get notified
         * if this fragment is still registered when the job finishes.
         */
        if (mCreateAfterScreenRotation) {
            generateMapList();
        }
    }

    private void generateMapList() {
        MapLoader.getInstance().generateMapArchives();
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMapImported(MapImportedEvent event) {
        Snackbar snackbar = Snackbar.make(mView, R.string.snack_msg_show_map_list, Snackbar.LENGTH_LONG);
        snackbar.setAction(R.string.ok_dialog, v -> mListener.onMapArchiveFragmentInteraction());
        snackbar.show();
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onRequestImportMapEvent(RequestImportMapEvent event) {
        String confirmImport = getContext().getString(R.string.confirm_import);
        Snackbar snackbar = Snackbar.make(getView(), confirmImport, Snackbar.LENGTH_LONG);
        snackbar.show();
    }

    /**
     * When this fragment is created for the first time, we wait the {@link android.support.v4.widget.DrawerLayout}
     * to close before generating the map list (to avoid a stutter). <br>
     */
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onDrawerClosed(DrawerClosedEvent event) {
        generateMapList();
    }

    /**
     * A {@link MapArchiveListUpdateEvent} is emitted from the {@link MapLoader} when the list of
     * map archives is updated.
     */
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMapArchiveListUpdate(MapArchiveListUpdateEvent event) {
        hideProgressBar();
    }

    @Override
    public void onStop() {
        EventBus.getDefault().unregister(this);
        if (mMapArchiveAdapter != null) {
            mMapArchiveAdapter.unSubscribeEventBus();
        }
        super.onStop();

        if (mRecyclerView == null) return;
        for (int i = 0; i < mRecyclerView.getAdapter().getItemCount(); i++) {
            MapArchiveViewHolder holder = (MapArchiveViewHolder) mRecyclerView.findViewHolderForAdapterPosition(i);
            if (holder != null) {
                holder.unSubscribe();
            }
        }
    }

    private void hideProgressBar() {
        mView.findViewById(R.id.progress_import_frgmt).setVisibility(View.GONE);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        outState.putBoolean(CREATE_FROM_SCREEN_ROTATE, true);
    }

    public interface OnMapArchiveFragmentInteractionListener {
        void onMapArchiveFragmentInteraction();
    }
}

File: app/src/main/java/com/peterlaurence/trekme/ui/mapimport/MapImportFragment.kt
package com.peterlaurence.trekme.ui.mapimport

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.MimeTypeMap
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.documentfile.provider.DocumentFile
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import com.peterlaurence.trekme.R
import com.peterlaurence.trekme.databinding.FragmentMapImportBinding
import com.peterlaurence.trekme.util.RecyclerItemClickListener
import com.peterlaurence.trekme.viewmodel.mapimport.MapImportViewModel
import com.peterlaurence.trekme.viewmodel.mapimport.UnzipMapImportedEvent
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Displays the list of maps archives available for import.
 * Leverages the Storage Access Framework (SAF) to get the list of [DocumentFile]s from a directory
 * selected by the user. The SAF guarantees that the application is granted access to that folder.
 * Using ContentResolver, an InputStream is created from a [DocumentFile] when the user selects it
 * press the extract button. Then, the view-model is involved for the rest of the process.
 *
 * @author P.Laurence on 08/06/16 -- Converted to Kotlin on 18/01/19
 */
@AndroidEntryPoint
class MapImportFragment : Fragment() {
    private var _binding: FragmentMapImportBinding? = null // backing field
    private val binding get() = _binding!!

    private var mapArchiveAdapter: MapArchiveAdapter? = null
    private var fabEnabled = false
    private lateinit var data: List<MapImportViewModel.ItemData>
    private var itemSelected: MapImportViewModel.ItemData? = null
    private val viewModel: MapImportViewModel by viewModels()

    private val mapImportLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == Activity.RESULT_OK) {
            result.data?.data?.let { uri ->
                lifecycleScope.launch {
                    showProgressBar()
                    binding.buttonImport.isEnabled = false
                    val zipDocs = listZipDocs(uri)
                    viewModel.updateUriList(zipDocs)
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        viewModel.itemLiveData.observe(this) {
            it?.let { mapArchiveList ->
                this.data = mapArchiveList
                mapArchiveAdapter?.setMapArchiveList(mapArchiveList)
                binding.progressListUris.visibility = View.GONE
                binding.welcomePanel.visibility = View.GONE
                binding.archiveListPanel.visibility = View.VISIBLE
            }
        }

        viewModel.unzipEvents.observe(this) {
            it?.let { event ->
                val itemData = data.firstOrNull { item ->
                    item.id == event.itemId
                } ?: return@let

                mapArchiveAdapter?.setUnzipEventForItem(itemData, event)

                if (event is UnzipMapImportedEvent) {
                    onMapImported()
                }
            }
        }

        setHasOptionsMenu(false)
    }

    override fun onCreateView(
            inflater: LayoutInflater, container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View {
        /* The action bar isn't managed by Compose */
        (requireActivity() as AppCompatActivity).supportActionBar?.apply {
            show()
            title = getString(R.string.import_title)
        }

        _binding = FragmentMapImportBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
        mapArchiveAdapter = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.buttonImport.setOnClickListener {
            val intent = Intent(Intent.ACTION_OPEN_DOCUMENT_TREE)
            mapImportLauncher.launch(intent)
        }

        val recyclerViewMapImport = binding.recyclerViewMapImport
        recyclerViewMapImport.setHasFixedSize(false)

        val llm = LinearLayoutManager(context)
        recyclerViewMapImport.layoutManager = llm

        mapArchiveAdapter = MapArchiveAdapter()
        recyclerViewMapImport.adapter = mapArchiveAdapter

        recyclerViewMapImport.addOnItemTouchListener(
            RecyclerItemClickListener(this.context,
                recyclerViewMapImport,
                object :
                    RecyclerItemClickListener.OnItemClickListener {

                    override fun onItemClick(view: View, position: Int) {
                        binding.fab.activate()
                        singleSelect(position)
                    }

                    override fun onItemLongClick(view: View?, position: Int) {
                        // no-op
                    }
                })
        )

        /* Item decoration : divider */
        val dividerItemDecoration = DividerItemDecoration(
                view.context,
                DividerItemDecoration.VERTICAL
        )
        val divider = ContextCompat.getDrawable(view.context, R.drawable.divider)
        if (divider != null) {
            dividerItemDecoration.setDrawable(divider)
        }
        recyclerViewMapImport.addItemDecoration(dividerItemDecoration)
    }

    private suspend fun listZipDocs(uri: Uri): List<DocumentFile> = withContext(Dispatchers.IO) {
        val context = context ?: return@withContext listOf<DocumentFile>()
        val docFile = DocumentFile.fromTreeUri(context, uri)
                ?: return@withContext listOf<DocumentFile>()
        if (docFile.isDirectory) {
            val zipDocs = docFile.listFiles().filter {
                MimeTypeMap.getSingleton().getExtensionFromMimeType(it.type) == "zip"
            }

            zipDocs
        } else listOf()
    }

    private fun singleSelect(position: Int) {
        /* Update adapter to reflect selection */
        mapArchiveAdapter?.setSelectedPosition(position)
        mapArchiveAdapter?.notifyDataSetChanged()

        /* Keep a reference on the selected archive */
        itemSelected = data[position]
    }

    private fun FloatingActionButton.activate() {
        val fab = binding.fab
        if (!fabEnabled) {
            fab.isEnabled = true
            fab.drawable.mutate().setTint(resources.getColor(R.color.colorWhite, null))
            fab.background.setTint(resources.getColor(R.color.colorAccent, null))

            fab.setOnClickListener {
                itemSelected?.let { item ->
                    viewModel.unarchiveAsync(item)
                }
            }
        }
    }

    private fun onMapImported() {
        val view = view ?: return
        val snackbar = Snackbar.make(view, R.string.snack_msg_show_map_list, Snackbar.LENGTH_LONG)
        snackbar.setAction(R.string.ok_dialog) {
            runCatching {
                findNavController().navigate(R.id.mapListFragment)
            }
        }
        snackbar.show()
    }

    private fun showProgressBar() {
        binding.progressListUris.visibility = View.VISIBLE
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        outState.putBoolean(CREATE_FROM_SCREEN_ROTATE, true)
    }

    companion object {
        private const val CREATE_FROM_SCREEN_ROTATE = "create"
    }
}
